components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenProcessor.java

```diff
@@ -0,0 +1,719 @@
+                LOG.debug("Error occurred while logging operation execution results.", e);
+            }
+        }
+    }
+
+    @Override
+    public ActionExecutionStatus processErrorResponse(ActionType actionType, Map<String, Object> map, Event event,
+                                                      ActionInvocationErrorResponse actionInvocationErrorResponse) {
+
+        //todo: need to implement to process the error so that if a processable error is received
+        // it is communicated to the client.
+        // we will look into this as we go along with other extension types validating the way to model this.
+        return null;
+    }
+
+    private void updateTokenMessageContext(OAuthTokenReqMessageContext tokenMessageContext,
+                                           AccessToken responseAccessToken) {
+
+        tokenMessageContext.setScope(responseAccessToken.getScopes().toArray(new String[0]));
+
+        String expiresInClaimName = CLAIMS_PATH_PREFIX + AccessToken.ClaimNames.EXPIRES_IN.getName();
+        responseAccessToken.getClaims().stream()
+                .filter(claim -> expiresInClaimName.equals(claim.getName()))
+                .findFirst()
+                .map(claim -> Long.parseLong(claim.getValue().toString()) * 1000)
+                .ifPresent(tokenMessageContext::setValidityPeriod);
+
+        Optional.ofNullable(responseAccessToken.getClaim(AccessToken.ClaimNames.AUD.getName()))
+                .map(AccessToken.Claim::getValue)
+                .ifPresent(value -> {
+                    List<String> audienceList;
+                    if (value instanceof List) {
+                        audienceList = (List<String>) value;
+                    } else {
+                        audienceList = Collections.emptyList();
+                    }
+                    tokenMessageContext.setAudiences(audienceList);
+                });
+
+        Map<String, Object> customClaims = new HashMap<>();
+        for (AccessToken.Claim claim : responseAccessToken.getClaims()) {
+            if (!AccessToken.ClaimNames.contains(claim.getName())) {
+                customClaims.put(claim.getName(), claim.getValue());
+            }
+        }
+        tokenMessageContext.setAdditionalAccessTokenClaims(customClaims);
+
+        tokenMessageContext.setPreIssueAccessTokenActionsExecuted(true);
+    }
+
+    private OperationExecutionResult handleAddOperation(PerformableOperation operation, AccessToken requestAccessToken,
+                                                        AccessToken.Builder responseAccessToken) {
+
+        if (operation.getPath().startsWith(SCOPE_PATH_PREFIX)) {
+            return addScope(operation, responseAccessToken);
+        } else if (operation.getPath().startsWith(CLAIMS_PATH_PREFIX)) {
+            return addClaim(operation, requestAccessToken, responseAccessToken);
+        }
+
+        return new OperationExecutionResult(operation, OperationExecutionResult.Status.FAILURE,
```

Comment:
So from this, we define that the access token can not have any first-level custom attributes other than scopes or claims. Won't there be requirements to add any first-level custom attributes?

Reply:
Well, claims allow to add first level claims to the access token.
Something like below.

`{
  "sub": "rBG40w5KDm8NJvD_JrPh1RLR7l4a",
  "aut": "APPLICATION",
  "iss": "https://localhost:9443/oauth2/token",
  "customArray": [
    "a",
    "b",
    "c"
  ],
  "client_id": "rBG40w5KDm8NJvD_JrPh1RLR7l4a",
  "aud": [
    "example.com",
    "https://example.com/resource"
  ],
  "nbf": 1721895331,
  "customSID": "12345",
  "azp": "rBG40w5KDm8NJvD_JrPh1RLR7l4a",
  "org_id": "10084a8d-113f-4211-a0d5-efe36b082211",
  "scope": "custom-scope-1",
  "exp": 1721898931,
  "org_name": "Super",
  "iat": 1721895331,
  "jti": "8d97e5ea-adca-4f70-bd93-fa22df12af12"
}`

Note the `customArray` and `customSID`